runtime/evaluate.cc

```diff
@@ -412,10 +417,24 @@ class evaluator {
     std::atomic<index_t> result = 0;
     std::size_t n = ceil_div(bounds.max - bounds.min + 1, step);
     context.reserve(op->sym.id + 1);
-    index_t old_value = context.set(op->sym, 0);
-    context.thread_pool->parallel_for(
+    context.config->thread_pool->parallel_for(
         n,
-        [context = this->context, step, min = bounds.min, op, &result](index_t i) mutable {
+        [parent_context = &context, step, min = bounds.min, op, &result](index_t i) mutable {
+          eval_context context;
+          if (const let_stmt* closure = is_closure(op->body)) {
+            // The body is a closure, so we know exactly which symbols we need to copy to the new local context.
+            context.reserve(parent_context->size());
+            context.config = parent_context->config;
+
+            // Assume that this let_stmt is a closure for this loop. We'll evaluate the values using the parent context,
+            // but assign them to our local context.
+            for (const std::pair<var, expr>& i : closure->lets) {
```

Comment:
A further optimization we maybe don't need, but I think we only need a new context for each worker once and  don't actually need it for each iteration of the loop.

Reply:
I was thinking about this too, but it's not easy to do. I think we need to move a lot of the logic for `parallel_for` here to do it.

I think we should do that for sure, but in a separate PR.